### ISSUE TITLE
add kaiju

### ISSUE THREAD

**Issue by pmenzel** [usegalaxy-eu/usegalaxy-eu-tools#690]
Hi!

I was wondering if it is possible to make [kaiju](https://kaiju.binf.ku.dk/) available for public usage on usegalaxy.eu.  
The current public web server for kaiju is always overloaded and will likely be out of service soon, so it would be very nice to make the program available at usegalaxy.eu. Similar tools like kraken2 are already there.

The main computational cost is the high memory requirement, depending on the used reference database, which might be prohibitive.
See [the kaiju download page](https://bioinformatics-centre.github.io/kaiju/downloads.html) for the memory requirements, they range from 49GB to 204GB (as of June 2023) from the smallest to the most comprehensive reference database.

If it is possible to add kaiju to the service, @pvanheus suggested to add it to the list of tools at https://github.com/galaxyproject/tools-iuc, which I could try to do, probably requiring some help.

Thanks!

**Comment by pvanheus** [comment]
@pmenzel as mentioned on Slack, first there needs to be a Galaxy wrapper for `kaiju` and also a data manager to manage its databases. Input from the usegalaxy.eu admins on those resource requirements would be useful, though, to get some idea of whether it could be accommodated if a wrapper was available.

**Comment by bgruening** [comment]
@pmenzel we can help you getting this tools into Galaxy. Do you know if the tool and its database will be maintained when they shutdown their server? Maybe they are interested to help you as well?

Here are a few links to checkout for Galaxy tools dev:

A few links and useful literature for Galaxy tool and workflow development.

1. please use Visual Studio Code with this extension: https://github.com/galaxyproject/galaxy-language-server
2. this is the Galaxy SDK with a lot of documentation: https://planemo.readthedocs.io/
3. there is also https://docs.galaxyproject.org/en/latest/dev/schema.html \<-- which is already integrated into the VSC extension above, but it might be nice to read, browse, search
4. there is also the Best practice guide: https://galaxy-iuc-standards.readthedocs.io \<-- which is already integrated into the VSC extension above, but it might be nice to read, browse, search

A complete (4h) tutorial, with everything you need is avaiable in the planeom docs. It will save you a lot of time later, please do this tutorial: https://planemo.readthedocs.io/en/latest/writing.html

If you want to read a publication about planemo have a look here: https://genome.cshlp.org/content/33/2/261.long

Last but not least there is this short blog post from David, showing the 3 major steps that you need to do if you want to get your tool into the European Galaxy server: https://usegalaxy-eu.github.io/posts/2020/08/22/three-steps-to-galaxify-your-tool/

**Comment by pmenzel** [comment]
@pvanheus Yes, exactly. Given the high resource requirements, I would need to know if it is at all possible, before starting to make a wrapper.. @bgruening and other admins, what do you think? 

@bgruening I am the only maintainer and I plan to keep updating the downloadable reference databases once per year as before.

**Comment by bgruening** [comment]
Ah I see! :)
Yes, we can run this tool on our infrastructure. If you could help us predict the memory requirement of a job that would help as well.

Thanks, let us know if you need any help.

**Comment by pmenzel** [comment]
The memory requirement comes from loading the reference index and does not depend on the size of input fastq/fasta files, so it is easy to predict. :) It currently ranges from 49GB to 204GB depending on the reference database (numbers from June 2023).

However, as sequence databases grow, these numbers will continue to increase and might well be up to 230GB in this year (I will make the databases for 2024 in the summer again).

From the CPU perspective, 10 parallel threads are enough for the program to chuck along.

**Comment by bgruening** [comment]
That all looks ok to me!

**Comment by bgruening** [comment]
@pmenzel do you need any help?

**Comment by pmenzel** [comment]
@bgruening I didn't get around to delve into this issue, unfortunately. If the open issue is bothering, we can also close it for now and I will comment again, once I started working on it.

**Comment by bgruening** [comment]
its not bothering, just keep us updated :) Thanks a lot!

**Comment by mycojon** [comment]
I would be VERY interested to see the KAIJU program on the Galaxay EU site.   I find that it can classify species that aren't identified by nucleotide mapping.  So if I question results, I can run them on KAIJU and compare results.  If I find a species on KAIJU, I can usually find it in my sequences.

I have had only two issues regarding the databases used by KAIJU, and it's the same issue on Kraken2, etc.   I have three species in the samples I'm running that aren't represented.

Mycobacterium - (Mycobacterium 1100029.7)   It's not recognized by Kraken2, as it's not in the database.  KAIJU catches it as Tuberculosis as I believe one of it's genes is the same as Tuberculosis.

Plasmodium (Plasmodium Ovale Wallikeri and Plasmodium Ovale Curtisi) BOTH of these are human pathogens, yet almost none of the databases available have it in them.  The NCBI has it, so it shows up in the KAIJU results, but it's not in the Kraken2 databases so it doesn't even recognize it as Plasmodium.   However a new reference was recently published POW222 (Plasmodium Ovale Wallikeri), and POC221 (Plasmodium Ovale Curtisi).

I was using KAIJU through Kbase, but it's use on Galaxy would be much better.

Thanks!!

**Comment by bernt-matthias** [comment]
Had a few minutes and started a PR here https://github.com/galaxyproject/tools-iuc/pull/6908. So far a really early state. 

@pmenzel @mycojon could someone come up with a small reference database? At IUC we have a maximum file size of 1MB. We would need something supporting a successful run. It does not need to be meaningful, but should produce some output.

**Comment by pmenzel** [comment]
@bernt-matthias Oh wow, thank you!

Do you really mean 1MB or 1GB?

The NCBI taxonomy files included in the downloadable kaiju index are already a few hundred MB. The smallest kaiju DB so far is the one for viruses, see [here](https://bioinformatics-centre.github.io/kaiju/downloads.html).

**Comment by bernt-matthias** [comment]
> Do you really mean 1MB or 1GB?

Really 1MB :) IUC contains tool wrappers for a few hundred packages, so we need to have an eye on test data size.

> The NCBI taxonomy files included in the downloadable kaiju index are already a few hundred MB. The smallest kaiju DB so far is the one for viruses, see [here](https://bioinformatics-centre.github.io/kaiju/downloads.html).

Maybe we can create a mock datasets consisting of a few viruses? Removing the "irrelevant" parts from the nodes and names dmp files should be easy.

**Comment by bernt-matthias** [comment]
That said, alternatively we can also "easily" setup a separate Galaxy tool repo for the kaiju Galaxy tool that will have the same CI as the IUC repo. There we could add custom reference data at will. 

We have a [template repo](https://github.com/galaxyproject/galaxy-tool-repository-template) for this purpose.

**Comment by mycojon** [comment]
Hi Peter,
One option that I think would be helpful, at least for issues I have run into.  Kaiju output shows counts associated with Taxon identified.   Which works fine in most cases, but I have had some issues.

It would be EXREMELY helpful if there was an option to download the read pairs in fastq for each of Taxon identified.  For example

In my blood samples I'm working with, KAIJU a couple years ago reported Mycobacterium Leprae in very high concentration, another online platform OneCodex also identified Mycobacterium Leprae.  Sounds great, I have two different classifiers that tell me Leprae is present, so I should have high confidence.

The problem is that if I run KAIJU on the Kbase platform, my multiple samples that previously said Mycobacterium Leprae, NOW all say Mycobacterium Tuberculosis.  OneCodex recently added the option to download reads for an identified Taxon.

I did this for Mycobacterium Leprae, downloaded from OneCodex.

I can assemble the reads or use individual reads and BLAST them.  With standard Blastn, I will usually get either Human or Naegleria Fowleri NF001, if I specify only Mycobacterium species all of the reads come up as Mycobacterium Tuberculosis MTB-Oman-3214883, if I look at the Leprae all of the Oman-3214883 reads align with a short section Leprae that matches with a Blastx alignment.

In looking at the reference sequences for the identified Leprae and MTB-Oman, these were all sequenced in the last couple of years, which is likely why KAIJU now identifies them as Tuberculosis.  But my dilemma is that I still don't know what the reads are so that I can assemble them to get a definitive alignment.  I can't even just map all of the 76 MTB-Oman strains and do de novo, as there are numerous other species that have similar Tuberculosis genes (Naegleria Fowleri, and several unclassified Mycobacterium).

Northwest Hygienetics, Inc., / CFS
Jon Colman
29903 43rd Ave S
Auburn,  WA  98001
253-529-0294
________________________________

**Comment by pmenzel** [comment]
@bernt-matthias I made a tiny subset of nodes.dmp and names.dmp restricted to SARS-CoV-2 and a Kaiju index with only the SARS-CoV-2 proteins. See attached zip file. It also has a test query.fa file. A few KB in total.

You can run `kaiju -t sarscov2-nodes.dmp -f sarscov2.fmi -i query.fa -v -o out` and `out` should contain the line
```
C	testseq	2697049	1918	2697049,	YP_009724389.1,YP_009725295.1,	RSTNSRIKASMPTTIAKNTVKSVGKFCLEASFNYLKSPNFSKLINIIIWFLLLSVCLGSLIYSTAALGVLMSNLGMPSYCTGYREGYLNSTNVTIATYCTGSIPCSVCLSGLDSLDTYPSLETIQITISSFKWDLTAFGLVAEWFLAYILFTRFFYVLGLAAIMQLFFSYFAVHFISNSWLMWLIINLVQMAPISAMVRMYIFFASFYYVWKSYVHVVDGCNSSTCMMCYKRNRATRVECTTIVNGVRRSFYVYANGGKGFCKLHNWNCVNCDTFCAGSTFISDEVARDLSLQFKRPINPTDQSSYIVDSVTVKNGSIHLYFDKAGQKTYERHSLSHFVNLDNLRANNTKGSLPINVIVFDG,
```
Also `kaiju2table -t sarscov2-nodes.dmp -n sarscov2-names.dmp -o out.tsv -r species -p -e out` should give:
```
file	percent	reads	taxon_id	taxon_name
out	100.000000	1	2697049	Viruses;Riboviria;Orthornavirae;Pisuviricota;Pisoniviricetes;Nidovirales;Cornidovirineae;Coronaviridae;Orthocoronavirinae;Betacoronavirus;Sarbecovirus;Severe acute respiratory syndrome-related coronavirus;Severe acute respiratory syndrome coronavirus 2;
out	0.000000	0	NA	cannot be assigned to a (non-viral) species
out	0.000000	0	NA	unclassified

```



[kaiju-test-db.zip](https://github.com/user-attachments/files/19673753/kaiju-test-db.zip)

**Comment by bernt-matthias** [comment]
> [@bernt-matthias](https://github.com/bernt-matthias) That's really awesome. I will have a look at the help files and need to check the columns of those two programs, it has been some time :)

Wonderful

> Would it be possible to create a standard workflow for simple read classification that would chain together the common steps, e.g:
> 
> 1. run kaiju with default params on the input fasta / fastq files.
> 2. run kaiju2krona on the output, followed by ktImportText, so that a krona html file will be generated by default.
> 3. run kaiju2table with a fixed set of options

Sure really good idea. 

> Then I could make a list of commands that should be run in succession for this workflow.

Let me know what these commands are, then I can plan better.

**Comment by bernt-matthias** [comment]
The tools should be on usegalaxy.eu  soon https://github.com/usegalaxy-eu/usegalaxy-eu-tools/pull/850

**Comment by pmenzel** [comment]
@bernt-matthias  Sorry for the late response and thanks for all your work!!

The columns for kaijux and kaijup (sequence search without taxonomic classification) are similar to the normal kaiju output columns:

1. C or U, with C = a matching sequence was found or, U = not found
2. Query sequence name from input file (e.g. read name)
3. Database Sequence ID(s), e.g. accession number(s), of the sequence(s) with the best match length (MEM mode) or match score (Greedy mode)
4. Sequence(s) with best match length (MEM mode) or match score (Greedy mode). Column 4 is only shown in verbose mode using command line option -v.


For a standard workflow that could be executed:

Run kaiju with default params on the input fasta / fastq files.

```
kaiju .... -v -o kaiju_out.tsv
```

Then, run kaiju2krona on the output, followed by ktImportText, so that a krona html file will be generated by default:

```
kaiju2krona -u -n names.dmp -t nodes.dmp -i  kaiju_out.tsv -o kaiju_krona.tsv

ktImportText -o kaiju_krona.html kaiju_krona.tsv
```
`ktImportText` might create a folder called `kaiju_krona.html.files` that can be deleted. `ktImportText` is contained in the conda package [krona](https://bioconda.github.io/recipes/krona/README.html), if this helps.
  
Run kaiju2table:

```
kaiju2table -e -u -r genus -l superkingdom,phylum,class,order,family,genus -n names.dmp -t nodes.dmp -m 0.1 -o kaiju_report_genus.tsv kaiju_out.tsv

kaiju2table -e -u -r species -l superkingdom,phylum,class,order,family,genus,species -n names.dmp -t nodes.dmp -m 0.1 -o kaiju_report_species.tsv kaiju_out.tsv
```

**Comment by bernt-matthias** [comment]
kaiju should become available as soon as https://github.com/usegalaxy-eu/usegalaxy-eu-tools/pull/850 is merged. for reference data someone from the usegalaxy-eu has to run the data manager. 

For a proper setup it will be helpful if you could give us some resource estimates for kaiju, i.e. run time and memory.

@bgruening any estimation? 

In case you are new to Galaxy I could show you around.

**Comment by bgruening** [comment]
Automatic install will be running over the weekend, next Monday it should be availabe.

**Comment by pmenzel** [comment]
The memory requirements for the latest reference databases are listed in the [downloads page](https://bioinformatics-centre.github.io/kaiju/downloads.html). It's probably a good idea to add a couple of GBs, when enforcing a hard limit though.

The runtime requirements depend on the input file size, but should not be more than a few hours for a couple million sequencing reads.

What also comes to my mind is that there could be a warning message in the input file selection, that sequencing adapters should already be removed from the sequencing reads before analysing them with kaiju.

**Comment by bernt-matthias** [comment]
Tried to build the desired workflow. Works in principle. A few bugfixes will come soon: https://github.com/galaxyproject/tools-iuc/pull/6957

The test data in https://github.com/pmenzel/kaiju-testdata are 2 samples, or? 

Is there some paired test data (maybe also multiple samples) somewhere?

**Comment by pmenzel** [comment]
yes, the test data set are two single samples. If a paried-end test data set is needed, I think you could just use the same fastq file for arguments `-i` and `-j`.

**Comment by pmenzel** [comment]
Hm.. works fine for me using the file `sars-cov-2_1.fastq.gz` from the test data repo:
```
~/software/kaiju/bin/kaiju -t sarscov2-nodes.dmp -f sarscov2.fmi -i sars-cov-2_1.fastq.gz -j sars-cov-2_1.fastq.gz -v -o out
Parameters: 
  run mode: Greedy
  minimum match length: 11
  seed length: 7
  minimum blosum62 score for matches: 65
  minimum E-value: 0.01
  max number of mismatches within a match: 3
  input file 1: sars-cov-2_1.fastq.gz
  input file 2: sars-cov-2_1.fastq.gz
  output file: out
09:59:16 Reading database
 Reading taxonomic tree from file sarscov2-nodes.dmp
 Reading index from file sarscov2.fmi
09:59:16 Start classification using 1 threads.
09:59:20 Finished.

```

**Comment by bgruening** [comment]
Kaiju is installed on EU. I will close this one, please reopen if still needed.
